app/src/main/scala/org/alephium/explorer/persistence/queries/EventQueries.scala

```diff
@@ -0,0 +1,88 @@
+// (at your option) any later version.
+//
+// The library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the library. If not, see <http://www.gnu.org/licenses/>.
+
+package org.alephium.explorer.persistence.queries
+
+import slick.jdbc.{PositionedParameters, SetParameter, SQLActionBuilder}
+import slick.jdbc.PostgresProfile.api._
+
+import org.alephium.explorer.api.model._
+import org.alephium.explorer.persistence._
+import org.alephium.explorer.persistence.model.EventEntity
+import org.alephium.explorer.persistence.schema.CustomGetResult._
+import org.alephium.explorer.persistence.schema.CustomSetParameter._
+import org.alephium.explorer.util.SlickUtil._
+import org.alephium.protocol.model.TransactionId
+
+object EventQueries {
+
+  def insertEventsQuery(events: Iterable[EventEntity]): DBActionW[Int] = {
+    QuerySplitter.splitUpdates(rows = events, columnsPerRow = 7) { (events, placeholder) =>
+      val query =
+        s"""
+           |INSERT INTO events ("block_hash", "tx_hash", "contract_address", "input_address","block_timestamp", "event_index", "fields")
+           |VALUES $placeholder
+           |ON CONFLICT
+           | ON CONSTRAINT events_pk
+           | DO NOTHING
+           |""".stripMargin
+
+      val parameters: SetParameter[Unit] =
+        (_: Unit, params: PositionedParameters) =>
+          events foreach { event =>
+            params >> event.blockHash
+            params >> event.txHash
+            params >> event.contractAddress
+            params >> event.inputAddress
+            params >> event.timestamp
+            params >> event.eventIndex
+            params >> event.fields
+        }
+
+      //Return builder generated by Slick's string interpolation
+      SQLActionBuilder(
+        queryParts = query,
+        unitPConv  = parameters
+      ).asUpdate
+    }
+  }
+
+  def getEventsByTxIdQuery(txId: TransactionId): DBActionSR[EventEntity] =
+    sql"""
+      SELECT *
+      FROM events
+      WHERE tx_hash = $txId
+      ORDER BY event_index
+      """.asASE[EventEntity](eventGetResult)
+
+  def getEventsByContractAddressQuery(address: Address): DBActionSR[EventEntity] =
+    sql"""
+      SELECT *
+      FROM events
+      WHERE contract_address = $address
+      ORDER BY block_timestamp DESC
+      """.asASE[EventEntity](eventGetResult)
+
+  def getEventsByContractAndInputAddressQuery(contract: Address,
+                                              input: Address): DBActionSR[EventEntity] =
+    sql"""
+      SELECT *
+      FROM events
+      WHERE contract_address = $contract
+      AND input_address = $input
+      ORDER BY block_timestamp DESC
+      """.asASE[EventEntity](eventGetResult)
```

Comment:
Neat. Should there be indexes for these columns accessed here?
- `event_index`
- `block_timestamp`
- `input_address`